### ISSUE TITLE
some question about the loss code~

### ISSUE THREAD

**Issue by x2x5** [HyelinNAM/ContrastiveDenoisingScore#13]
Hi there! Thanks for sharing your great work and for releasing such elegant and easy-to-run code!

I’ve been going through your code and have a few questions:

1. Regarding the code in the red rectangle in the screenshot: Is it actually needed? It seems that the computed values are never used later.
![image](https://github.com/user-attachments/assets/7156f73b-1540-4d88-a674-7e60d5c54112)

2. In terms of the loss computation, from your code it appears that you do:
   ```
   optimizer.zero_grad()
   dds_loss.backward()
   cut_loss.backward()
   optimizer.step()
   ```
   Does this effectively mean the total loss is `dds_loss + cut_loss`? In other words, is it equivalent to:
   ```
   optimizer.zero_grad()
   loss = dds_loss + cut_loss
   loss.backward()
   optimizer.step()
   ```

**Comment by HyelinNAM** [comment]
Hi, Thank you for your interest!

First of all, I apologize for the lack of clarity. For the experiment, where I need to toggle certain losses, there ended up being some redundancy in the code.

1. That part is simply for obtaining the self-attention feature, so those values are not used in the way you might have understood.

2. The reason I implemented it this way is that, in the case of the DDS loss, it’s known that the UNet Jacobian gradient can be skipped. Unlike the cut loss, there’s no need for backpropagation through the UNet. To stay consistent with the original DDS implementation, I first backpropagate the DDS loss (without going through the UNet), and then separately backpropagate the cut loss (which does go through the UNet).

Hope this helps you!

Best,
Hyelin

**Comment by x2x5** [comment]
> Hi, Thank you for your interest!
> 
> First of all, I apologize for the lack of clarity. For the experiment, where I need to toggle certain losses, there ended up being some redundancy in the code.
> 
> 1. That part is simply for obtaining the self-attention feature, so those values are not used in the way you might have understood.
> 2. The reason I implemented it this way is that, in the case of the DDS loss, it’s known that the UNet Jacobian gradient can be skipped. Unlike the cut loss, there’s no need for backpropagation through the UNet. To stay consistent with the original DDS implementation, I first backpropagate the DDS loss (without going through the UNet), and then separately backpropagate the cut loss (which does go through the UNet).
> 
> Hope this helps you!
> 
> Best, Hyelin

I understood~ thank you very much for your so quick and detailed feedback~
